src/dotty/tools/dotc/typer/Namer.scala

```diff
@@ -395,32 +395,64 @@ class Namer { typer: Typer =>
   /** Create top-level symbols for statements and enter them into symbol table */
   def index(stats: List[Tree])(implicit ctx: Context): Context = {
 
+    val classDef  = mutable.Map[TypeName, TypeDef]()
+    val moduleDef = mutable.Map[TypeName, TypeDef]()
+
     /** Merge the definitions of a synthetic companion generated by a case class
      *  and the real companion, if both exist.
      */
     def mergeCompanionDefs() = {
-      val classDef = mutable.Map[TypeName, TypeDef]()
       for (cdef @ TypeDef(name, _) <- stats)
-        if (cdef.isClassDef) classDef(name) = cdef
-      for (mdef @ ModuleDef(name, _) <- stats)
+        if (cdef.isClassDef) {
+          classDef(name) = cdef
+          cdef.attachmentOrElse(ExpandedTree, cdef) match {
+            case Thicket(cls :: mval :: (mcls @ TypeDef(_, _: Template)) :: crest) =>
+              moduleDef(name) = mcls
+            case _ =>
+          }
+        }
+      for (mdef @ ModuleDef(name, _) <- stats if !mdef.mods.is(Flags.Package)) {
+        val typName = name.toTypeName
+        val Thicket(vdef :: (mcls @ TypeDef(_, impl: Template)) :: Nil) = mdef.attachment(ExpandedTree)
+        moduleDef(typName) = mcls
         classDef get name.toTypeName match {
           case Some(cdef) =>
-            val Thicket(vdef :: (mcls @ TypeDef(_, impl: Template)) :: Nil) = mdef.attachment(ExpandedTree)
             cdef.attachmentOrElse(ExpandedTree, cdef) match {
               case Thicket(cls :: mval :: TypeDef(_, compimpl: Template) :: crest) =>
                 val mcls1 = cpy.TypeDef(mcls)(
                     rhs = cpy.Template(impl)(body = compimpl.body ++ impl.body))
                 mdef.putAttachment(ExpandedTree, Thicket(vdef :: mcls1 :: Nil))
+                moduleDef(typName) = mcls1
                 cdef.putAttachment(ExpandedTree, Thicket(cls :: crest))
               case _ =>
             }
           case none =>
         }
+      }
+    }
```

Comment:
I think this is a bit on the complex side, in particuar I do not like the repetition of the complicated Thicket pattern matches. Can we experiment with having 4 Maps:

 `classDef`, `moduleDef` as before

     val syntheticCompanionDef: Map[TypeName, TypeDef]  // a map from case classes to synthetic companions
     val coreCaseClassDef: Map[TypeName, List[Tree]] // a map from case classes to the definitions w/o synthetic companions

Using these we could streamline the traversal of ModuleDefs. 

See what I mean?